### ISSUE TITLE
Wrong results for exhaustive search

### ISSUE THREAD

**Issue by lkdvos** [bsc-quantic/EinExprs.jl#37]
The new implementations of the exhaustive search seem to have some bugs in them, the following works for version 0.5.7, but no longer works on the master branch:
```julia
A = EinExpr([:A, :a, :b, :c], Dict(:A => 2, :a => 2, :b => 2, :c => 2))
B = EinExpr([:b, :d, :e, :f], Dict(:b => 2, :d => 2, :e => 2, :f => 2))
C = EinExpr([:a, :e, :g, :C], Dict(:a => 2, :e => 2, :g => 2, :C => 2))
D = EinExpr([:c, :h, :d, :i], Dict(:c => 2, :h => 2, :d => 2, :i => 2))
E = EinExpr([:f, :i, :g, :j], Dict(:f => 2, :i => 2, :g => 2, :j => 2))
F = EinExpr([:B, :h, :k, :l], Dict(:B => 2, :h => 2, :k => 2, :l => 2))
G = EinExpr([:j, :k, :l, :D], Dict(:j => 2, :k => 2, :l => 2, :D => 2))

expr = EinExpr([:A, :B, :C, :D], [A, B, C, D, E, F, G])
config = Exhaustive()
expr_opt = EinExprs.einexpr(config, expr);
```
(v0.5.7)
```julia-repl
julia> map(string, branches(expr_opt))
6-element Vector{String}:
 "Bhkl,jklD->BhjD"
 "chdi,Aabc->hdiAab"
 "aegC,figj->aeCfij"
 "bdef,aeCfij->bdaCij"
 "hdiAab,bdaCij->hACj"
 "BhjD,hACj->ABCD"
```
(master)
```julia-repl
julia> map(string, branches(expr_opt))
6-element Vector{String}:
 "Aabc,bdef->bdef"
 "bdef,aegC->egC"
 "chdi,figj->igj"
 "egC,igj->gj"
 "Bhkl,jklD->klD"
 "gj,klD->ABCD"
```
The first line is clearly not a valid contraction, as the indices dont add up.

**Comment by mofeing** [comment]
Funk, I'll check it out. Thx for checking.

**Comment by mofeing** [comment]
@lkdvos seems like the version 0.5.7 doesn't give the result you're providing. Your result has a cost of `896` FLOPs while version 0.5.7 gives a count of `1088`, which is slightly worse.

**Comment by mofeing** [comment]
Fix in fa2aa93efb36421611544eb1865a7bf3057d2fc8

Still the fact that is not finding the optimal path is annoying me.

**Comment by mofeing** [comment]
Okay, I'm on it. I'm convinced the correct result is 896. But how did you get the result in v0.5.7?

**Comment by lkdvos** [comment]
Sorry my bad, I was experimenting with the TensorOperations wrapper, which shows the cost, and that's the one that gave me that contraction. #38 should give you what I have, I rebased it on the current master, feel free to experiment with it!

**Comment by mofeing** [comment]
Bug fixed! The bug was here: https://github.com/bsc-quantic/EinExprs.jl/blob/30bffcd75558f4b75b2c4dce675ec551177f90fd/src/Optimizers/Exhaustive.jl#L36

But it should be `if length(path.args) == 1`. It is now fixed in 8b58f75.

Furthermore, this bug fix speeds up the execution time by x2, and allocations by x2-3, of your example above.
- Before: ~50 ms (1.9M allocations: ~80 MiB)
- After: 23.781 ms (786457 allocations: 47.69 MiB)

**Comment by lkdvos** [comment]
🎉🎉
